apps/system/js/sync_manager.js

```diff
@@ -0,0 +1,460 @@
+          if (reg.task === SYNC_TASK) {
+            return !!(existingReg = reg);
+          }
+        });
+
+        if (existingReg) {
+          this.debug('Sync request already exists');
+          return Promise.resolve();
+        }
+
+        this.debug('Registering sync request', location.href);
+
+        return navigator.sync.register(SYNC_TASK, {
+          // XXX Create settings for this.
+          // minInterval is in seconds.
+          minInterval: 120, //XXX just to test
+          oneShot: true,
+          wifiOnly: false,
+          data: null,
+          wakeUpPage: location.href
+        });
+      });
+    },
+
+    unregisterSyncRequest: function() {
+      navigator.sync.unregister(SYNC_TASK).catch(e => {
+        // A failure unregistering the sync task might be expected
+        // depending on the state we are (i.e. enabling). In any case,
+        // we show a warning, so a bug can be identified or discarded.
+        console.warning('Could not unregister sync task', e,
+                        Service.query('SyncStateMachine.state'));
+      });
+    },
+
+    getKeys: function() {
+      return new Promise((resolve, reject) => {
+        LazyLoader.load('js/fx_accounts_client.js', () => {
+          FxAccountsClient.getKeys(resolve, reject);
+        });
+      });
+    },
+
+    getAssertion: function() {
+      return new Promise((resolve, reject) => {
+        LazyLoader.load('js/fx_accounts_client.js', () => {
+          FxAccountsClient.getAssertion(null, resolve, reject);
+        });
+      });
+    },
+
+    connect: function() {
+      if (this._port) {
+        return Promise.resolve(this._port);
+      }
+
+      return new Promise((resolve, reject) => {
+        navigator.mozApps.getSelf().onsuccess = event => {
+          var app = event.target.result;
+          app.connect(SYNC_REQUEST_IAC_KEYWORD).then(ports => {
+            if (!ports || !ports.length) {
+              return reject();
+            }
+            this._port = ports[0];
+            resolve(this._port);
+          }).catch(error => {
+            console.error(error);
+            reject(error);
+          });
+        };
+      });
+    },
+
+    iacRequest: function(request) {
+      return new Promise((resolve, reject) => {
+        this.connect().then(port => {
+          LazyLoader.load('/shared/js/uuid.js', () => {
+            var id = uuid();
+            request.id = id;
+            port.postMessage(request);
+            port.onmessage = event => {
+              var message = event.data;
+              if (!message || (message.id != id)) {
+                return;
+              }
+              resolve(message);
+            };
+          });
+        });
+      });
+    },
+
+    doSync: function(assertion, keys, collections) {
+      this.debug('Syncing with', collections);
+      this.iacRequest({
+        assertion: assertion,
+        keys: keys,
+        collections: collections
```

Comment:
Also need the Syncto server URL here, as well as xClientState. And `keys == (kA, kB)`, right?